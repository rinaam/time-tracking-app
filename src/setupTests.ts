// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import oranizationMembershipsResponse from './tests/fixtures/organizationMemberships';
import servicesFixture from './tests/fixtures/services';
import timeEntries from './tests/fixtures/timeEntries';

jest.useFakeTimers().setSystemTime(new Date('2022-05-21'));

// nice article https://kentcdodds.com/blog/stop-mocking-fetch

async function mockFetch(url: string) {
  switch (url) {
    case 'https://api.productive.io/api/v2/organization_memberships': {
      return {
        ok: true,
        status: 200,
        json: async () => oranizationMembershipsResponse,
      };
    }
    case 'https://api.productive.io/api/v2/services': {
      return {
        ok: true,
        status: 200,
        json: async () => servicesFixture,
      };
    }
    case 'https://api.productive.io/api/v2/time_entries?filter[person_id]=271497&filter[after]=2022-05-21&filter[before]=2022-05-22': {
      return {
        ok: true,
        status: 200,
        json: async () => timeEntries,
      };
    }

    default: {
      throw new Error(`Unhandled request: ${url}`);
    }
  }
}

// eslint-disable-next-line import/prefer-default-export
export const mockDeleteRequestRefetch = (timeEntryId: string) => async () => ({
  ok: true,
  status: 200,
  json: async () => ({
    ...timeEntries,
    data: timeEntries.data.filter((entry) => entry.id !== timeEntryId),
  }),
});

beforeAll(() => jest.spyOn(window, 'fetch'));
beforeEach(() => (window.fetch as jest.Mock).mockImplementation(mockFetch));
