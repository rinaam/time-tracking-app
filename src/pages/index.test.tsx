import React from 'react';
import { render, waitFor, screen, within, fireEvent } from '@testing-library/react';
import HomePage from '.';
import { mockDeleteRequestRefetch } from '../setupTests';

describe('home page', () => {
  it('match snapshot with all data', async () => {
    const { asFragment } = render(<HomePage />);

    await waitFor(() => {
      expect(screen.getByTestId('time-entries-table')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should delete time entry', async () => {
    render(<HomePage />);

    await waitFor(() => {
      expect(screen.getByTestId('row-delete-14541918')).toBeInTheDocument();
    });

    const row = await screen.getByTestId('row-delete-14541918');

    const button = within(row).getByText('delete');

    (window.fetch as jest.Mock).mockImplementation(mockDeleteRequestRefetch('14541918'));

    fireEvent.click(button);

    await waitFor(() => {
      expect(screen.queryByTestId('row-delete-14541918')).not.toBeInTheDocument();
    });
  });
});
