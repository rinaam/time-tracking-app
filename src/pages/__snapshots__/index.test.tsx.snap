// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home page match snapshot with all data 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <form
      class="time-entry-wrapper"
      data-testid="time-entries-table"
    >
      <div
        class="section"
      >
        <div
          class="dropdown-wrapper"
        >
          <label
            class="dropdown-label"
          >
            Service
          </label>
          <div
            class="dropdown-container"
            data-testid="test-id-service"
          >
            <p />
          </div>
        </div>
        <div
          class="input-wrapper"
        >
          <label
            class="input-label"
            for="time"
          >
            Time
          </label>
          <input
            class="input-field"
            data-testid="test-id-time"
            id="time"
            name="time"
            type="number"
            value=""
          />
        </div>
      </div>
      <div
        class="text-wrapper"
      >
        <label
          class="text-label"
          for="note"
        >
          Note
        </label>
        <textarea
          class="text-field"
          data-testid="test-id-note"
          id="note"
          name="note"
        />
      </div>
      <button
        class="btn-wrapper btn-disabled"
        disabled=""
        type="submit"
      >
        Save
      </button>
    </form>
    <table
      class="table-wrapper"
    >
      <thead>
        <tr>
          <th>
            Actions
          </th>
          <th>
            Date
          </th>
          <th>
            Service
          </th>
          <th>
            Person
          </th>
          <th>
            Note
          </th>
          <th>
            Time
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-testid="row-delete-14541918"
        >
          <td>
            <button
              class="btn-wrapper"
              type="button"
            >
              delete
            </button>
          </td>
          <td>
            2022-05-21
          </td>
          <td>
            App design
          </td>
          <td>
            Marina  Savic
          </td>
          <td>
            <ul>
              <li>
                foo bar
              </li>
            </ul>
          </td>
          <td>
            100
          </td>
        </tr>
        <tr
          data-testid="row-delete-14541903"
        >
          <td>
            <button
              class="btn-wrapper"
              type="button"
            >
              delete
            </button>
          </td>
          <td>
            2022-05-21
          </td>
          <td>
            Design
          </td>
          <td>
            Marina  Savic
          </td>
          <td>
            <ul>
              <li>
                2
              </li>
            </ul>
          </td>
          <td>
            60
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
